pkg/core/mpt/batch.go

```diff
@@ -0,0 +1,253 @@
+// and returns amount of the elements processed.
+// However each element is being put atomically, so Trie is always in a valid state.
+// It is used mostly after the block processing to update MPT and error is not expected.
+func (t *Trie) PutBatch(b Batch) (int, error) {
+	r, n, err := t.putBatch(b.kv)
+	t.root = r
+	return n, err
+}
+
+func (t *Trie) putBatch(kv []keyValue) (Node, int, error) {
+	return t.putBatchIntoNode(t.root, kv)
+}
+
+func (t *Trie) putBatchIntoNode(curr Node, kv []keyValue) (Node, int, error) {
+	switch n := curr.(type) {
+	case *LeafNode:
+		return t.putBatchIntoLeaf(n, kv)
+	case *BranchNode:
+		return t.putBatchIntoBranch(n, kv)
+	case *ExtensionNode:
+		return t.putBatchIntoExtension(n, kv)
+	case *HashNode:
+		return t.putBatchIntoHash(n, kv)
+	default:
+		panic("invalid MPT node type")
+	}
+}
+
+func (t *Trie) putBatchIntoLeaf(curr *LeafNode, kv []keyValue) (Node, int, error) {
+	t.removeRef(curr.Hash(), curr.Bytes())
+	return t.newSubTrieMany(nil, kv, curr.value)
+}
+
+func (t *Trie) putBatchIntoBranch(curr *BranchNode, kv []keyValue) (Node, int, error) {
+	return t.addToBranch(curr, kv, true)
+}
+
+func (t *Trie) mergeExtension(prefix []byte, sub Node) Node {
+	switch sn := sub.(type) {
+	case *ExtensionNode:
+		t.removeRef(sn.Hash(), sn.bytes)
+		sn.key = append(prefix, sn.key...)
+		sn.invalidateCache()
+		t.addRef(sn.Hash(), sn.bytes)
+		return sn
+	case *HashNode:
+		return sn
+	default:
+		if len(prefix) != 0 {
+			e := NewExtensionNode(prefix, sub)
+			t.addRef(e.Hash(), e.bytes)
+			return e
+		}
+		return sub
+	}
+}
+
+func (t *Trie) putBatchIntoExtension(curr *ExtensionNode, kv []keyValue) (Node, int, error) {
+	t.removeRef(curr.Hash(), curr.bytes)
+
+	common := lcpMany(kv)
+	pref := lcp(common, curr.key)
+	if len(pref) == len(curr.key) {
+		// Extension must be split into new nodes.
+		stripPrefix(len(curr.key), kv)
+		sub, n, err := t.putBatchIntoNode(curr.next, kv)
+		return t.mergeExtension(pref, sub), n, err
+	}
+
+	if len(pref) != 0 {
+		stripPrefix(len(pref), kv)
+		sub, n, err := t.putBatchIntoExtensionNoPrefix(curr.key[len(pref):], curr.next, kv)
+		return t.mergeExtension(pref, sub), n, err
+	}
+	return t.putBatchIntoExtensionNoPrefix(curr.key, curr.next, kv)
+}
+
+func (t *Trie) putBatchIntoExtensionNoPrefix(key []byte, next Node, kv []keyValue) (Node, int, error) {
+	b := NewBranchNode()
+	if len(key) > 1 {
+		b.Children[key[0]] = t.newSubTrie(key[1:], next, false)
+	} else {
+		b.Children[key[0]] = next
+	}
+	return t.addToBranch(b, kv, false)
+}
+
+func isEmpty(n Node) bool {
+	hn, ok := n.(*HashNode)
+	return ok && hn.IsEmpty()
+}
+
+// addToBranch puts items into the branch node assuming b is not yet in trie.
+func (t *Trie) addToBranch(b *BranchNode, kv []keyValue, inTrie bool) (Node, int, error) {
+	if inTrie {
+		t.removeRef(b.Hash(), b.bytes)
+	}
+	n, err := t.iterateBatch(kv, func(c byte, kv []keyValue) (int, error) {
+		child, n, err := t.putBatchIntoNode(b.Children[c], kv)
```

Comment:
Can we only pass a part of `kv` belonging to this child here?

Reply:
`iterateBatch` does exactly this, `kv` is shadowed here.